### PR TITLE
Fix thread-locals to be visible by Boehm

### PR DESCRIPTION
**Pull request by jacob-hughes**
Previously, thread-locals are invisible to Boehm. In other words, a pointer to a heap object stored solely inside a thread-local would cause the GC to free the object too early.

In order to do this, we must disable compiler-based thread local implementation. This is a shame because it's faster than the POSIX thread library impl but it places thread locals in a table in a segment in the binary which Boehm can't scan. It might be possible to fix this in the future, but for now we fallback to using the POSIX thread implementation because there exists a fix for this.

We introduce a global thread-local rootset, which holds pointers to thread-local values which Boehm can scan and use.

### CODE REVIEW

File: tests/ui/runtime/gc/thread_local.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+fn main() {
+    let t2 = std::thread::spawn(do_stuff_with_tls);
+
+    // Wait a little bit of time for the t2 to initialise thread-locals.
+    thread::sleep(time::Duration::from_millis(10));
+
+    GcAllocator::force_gc();
+
+    let _ = t2.join().unwrap();
+
```
Comment (ltratt) [inline tests/ui/runtime/gc/thread_local.rs:70]: I have a profound comment on this PR: extraneous blank line.

Comment (jacob-hughes) [reply tests/ui/runtime/gc/thread_local.rs]: Fixed here b9727c357795ccdcc7417ea3364dd1e8fe0e9d54

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by ltratt** [comment]
Hmm, weird. Let me try again.

**Comment by jacob-hughes** [comment]
Closed in favour of https://github.com/softdevteam/alloy/pull/106
